network-ntt/src/server.rs

```diff
@@ -0,0 +1,313 @@
+// because tokio-protocol is not generalized yet.
+// This should be fixed and then these types should go.
+use cardano::{
+    block::{HeaderHash},
+    tx::{TxAux},
+};
+
+/// Network transport server configuration.
+pub struct Config {
+    listen: Vec<SocketAddr>,
+    connect: Vec<SocketAddr>,
+}
+
+/// Internal structure of network transport node.
+#[derive(Clone)]
+pub struct Server<N> {
+    node: N,
+}
+
+/// Run NTT server.
+pub fn run<N: 'static>(config: Config, node: N)
+where
+    N: std::marker::Sync + std::marker::Send + server::Node + Clone + Copy,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::TipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::GetHeadersFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::StreamBlocksToTipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::TransactionService as network_core::server::transaction::TransactionService>::ProposeTransactionsFuture: std::marker::Send,
+{
+
+        let listener =
+            stream::iter_ok(config.listen)
+                .for_each(move |sockaddr| {
+                    let server = Server {
+                          node: node.clone(),
+                    };
+                    run_listen_socket(sockaddr, server) 
+                });
+        let connections =
+            stream::iter_ok(config.connect)
+                .for_each(move |sockaddr| {
+                let server = Server {
+                      node: node.clone(),
+                };
+                run_connect_socket(sockaddr, server)
+            });
+        tokio::run(connections.join(listener).map(|_| ()));
+}
+
+
+/// Run a server that will listen on a specific sockets
+/// and accept all incomming connections.
+/// Server maintains all of the incomming connection and
+/// `run_connection` is spawned on each of those connections.
+pub fn run_listen_socket<N: 'static>(sockaddr: SocketAddr, node: Server<N>) -> tokio::executor::Spawn
+where
+    N: server::Node + Clone,
+    N: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::TipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::GetHeadersFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::StreamBlocksToTipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::TransactionService as network_core::server::transaction::TransactionService>::ProposeTransactionsFuture: std::marker::Send,
+{
+    let server = TcpListener::bind(&sockaddr)
+        .expect("Can't bind socket")
+        .incoming()
+        .map_err(move |_err| ())
+        .for_each(move |stream| {
```

Comment:
I recommend separating out the listener stream creation to a helper function,
like done in `network-grpc`. The idea is for the application to be able to examine incoming connections as bare socket objects and decide if it wants to create protocol connections around them.

I will add another PR where the transport connection type is generalized in network-grpc, allowing TCP and Unix sockets. This is where such design also comes helpful.